ApollosProject/apollos-shovel
# Crossroads TV Backfill and Sync (via Contentful)

### About

This PR adds the backfill and sync for the Crossroads TV content.

### Backfill:
<img width="775" alt="Screen Shot 2022-02-07 at 12 10 19 PM" src="https://user-images.githubusercontent.com/34075293/152837167-f74e9d39-6a0d-4310-bea1-17c479569eee.png">

#### `create_categories`
In this step, we're creating 3 hard-coded categories for the Crossroads TV content: In-Room Messages, Messages, and Videos.

#### `fetch_and_save_entries`
In this step we retrieve entries from contentful and save them to postgres. At the same time, we'll save the asset associated with an entry, save an item to the media table for it's video URL, and link it to the correct category.

#### `fetch_and_delete_items`
Because the step above is specific for entries, we also have to subscribe to a `deletion` sync. This step handles entries and assets that were deleted from contentful.

#### `link_entries`
This step links parent content (collections or series) to their children. Each child in `content_items` is updated with a `parent_id`, and a row is added to the `content_item_connections` table.

### Sync:
<img width="553" alt="Screen Shot 2022-02-07 at 12 20 03 PM" src="https://user-images.githubusercontent.com/34075293/152838986-51f51d9d-3c45-4d0b-8a7d-433c1fe7b657.png">

The sync process does not include `create_categories`, but is otherwise the same as the backfill. We're storing the sync tokens in redis, and only content added or edited since the last sync token is synced.

### Testing

1. Start up Airflow, and filter the dags list by `crossroads_tv`
2. Unpause the `crossroads_tv_backfill_contentful_content_item_dag`. Wait for it to complete, it usually takes >20 minutes. (If you don't want to wait for it to complete, you can check out the crossroads_tv staging postgres DB and look at the data from the last time I ran the DAG.
3. Unpause the `crossroads_tv_contentful_content_item_dag`. Wait for it to finish. It should only take a few minutes.
4. Check out all the data in postgres 😅

### Screenshots
 

|Backfill|Sync|
|---|---|
|<img width="377" alt="Screen Shot 2022-02-07 at 12 23 34 PM" src="https://user-images.githubusercontent.com/34075293/152839565-1f684f4f-6004-461f-ae54-3f7ebec002b5.png">|<img width="436" alt="Screen Shot 2022-02-07 at 12 23 22 PM" src="https://user-images.githubusercontent.com/34075293/152839583-0b633def-fa64-4956-9002-9a1679d7da1e.png">|
